web/modules/custom/pulses_exchange_rate/src/CurrencyDataProvider.php

```diff
@@ -0,0 +1,173 @@
+<?php
+
+namespace Drupal\pulses_exchange_rate;
+
+use Drupal\Component\Datetime\TimeInterface;
+use Drupal\Core\Cache\CacheBackendInterface;
+use Drupal\Core\Config\ConfigFactoryInterface;
+use Drupal\pulses_exchange_rate\Form\ExchangeApiKey;
+use GuzzleHttp\ClientInterface;
+
+/**
+ * Object for get data from api.
+ */
+class CurrencyDataProvider implements CurrencyDataProviderInterface {
+
+  /**
+   * Client instance.
+   *
+   * @var \GuzzleHttp\ClientInterface
+   */
+  public $client;
+
+  /**
+   * Http header.
+   *
+   * @var object
+   */
+  private $header;
+
+  /**
+   * Final array with different nested.
+   *
+   * @var array
+   */
+  public $data = [];
+
+  /**
+   * Instance of ConfigFactory.
+   *
+   * @var \Drupal\Core\Config\ConfigFactoryInterface
+   */
+  private $configFactory;
+
+  /**
+   * Instance of CacheBackend.
+   *
+   * @var \Drupal\Core\Cache\CacheBackendInterface
+   */
+  private $cache;
+
+  /**
+   * Instance of Time.
+   *
+   * @var \Drupal\Component\Datetime\TimeInterface
+   */
+  private $dateTime;
+
+  /**
+   * Instance of Client class.
+   *
+   * @param \GuzzleHttp\ClientInterface $http_client
+   *   Client object.
+   * @param \Drupal\Core\Config\ConfigFactoryInterface $config_factory
+   *   ConfigFactory object.
+   * @param \Drupal\Core\Cache\CacheBackendInterface $cache
+   *   CacheBackend object.
+   * @param \Drupal\Component\Datetime\TimeInterface $date_time
+   *   TimeInterface object.
+   */
+  public function __construct(ClientInterface $http_client, ConfigFactoryInterface $config_factory, CacheBackendInterface $cache, TimeInterface $date_time) {
+    $this->client = $http_client;
+    $this->configFactory = $config_factory;
+    $this->cache = $cache;
+    $this->dateTime = $date_time;
+  }
+
+  /**
+   * Request data by URL & write value to object variable (header).
+   *
+   * Sort array with data after decode & write value to object variable (data).
+   *
+   * @param string $url
+   *   Url of API page.
+   * @param string $api_key
+   *   Single parameter for API Url.
+   * @param bool $nested
+   *   Describe how to sort array with data after decode.
+   */
+  public function getResponse($url, $api_key, $nested) {
+    $cid = 'pulses_exchange_rate: nested_data_' . var_export($nested, TRUE);
```

Comment:
What will be when the API key has been changed? Correct, the cache will return the data for the previous API key. You need to add the API key to the cache ID. Also, will be good to make a separate function to prepare the cache ID to not duplicate the code in the file. E.g. you can replace this code to:
```suggestion
    $this->apiKey = $api_key;
    $cid = $this->makeCacheId($nested);
```
and create a separate function 
```
  private function makeCacheId($nested) {
    return 'pulses_exchange_rate:' . $this->apiKey . ':nested_data_' . var_export($nested, TRUE);
  }
```